src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -521,103 +524,90 @@ internal unsafe SocketError DoOperationReceiveFromMultiBuffer(SafeSocketHandle h
-            }
-
-            // Create and pin an appropriately sized control buffer if none already
-            IPAddress? ipAddress = (_socketAddress!.Family == AddressFamily.InterNetworkV6 ? _socketAddress.GetIPAddress() : null);
-            bool ipv4 = (_currentSocket!.AddressFamily == AddressFamily.InterNetwork || (ipAddress != null && ipAddress.IsIPv4MappedToIPv6)); // DualMode
-            bool ipv6 = _currentSocket.AddressFamily == AddressFamily.InterNetworkV6;
-
-            if (ipv6 && (_controlBufferPinned == null || _controlBufferPinned.Length != sizeof(Interop.Winsock.ControlDataIPv6)))
-            {
-                _controlBufferPinned = GC.AllocateUninitializedArray<byte>(sizeof(Interop.Winsock.ControlDataIPv6), pinned: true);
-            }
-            else if (ipv4 && (_controlBufferPinned == null || _controlBufferPinned.Length != sizeof(Interop.Winsock.ControlData)))
-            {
-                _controlBufferPinned = GC.AllocateUninitializedArray<byte>(sizeof(Interop.Winsock.ControlData), pinned: true);
-            }
-
-            // If single buffer we need a single element WSABuffer.
-            WSABuffer[] wsaRecvMsgWSABufferArray;
-            uint wsaRecvMsgWSABufferCount;
             if (_bufferList == null)
             {
-                if (_wsaRecvMsgWSABufferArrayPinned == null)
-                {
-                    _wsaRecvMsgWSABufferArrayPinned = GC.AllocateUninitializedArray<WSABuffer>(1, pinned: true);
-                }
-
                 fixed (byte* bufferPtr = &MemoryMarshal.GetReference(_buffer.Span))
                 {
                     Debug.Assert(_singleBufferHandleState == SingleBufferHandleState.None);
                     _singleBufferHandleState = SingleBufferHandleState.InProcess;
 
-                    _wsaRecvMsgWSABufferArrayPinned[0].Pointer = (IntPtr)bufferPtr + _offset;
-                    _wsaRecvMsgWSABufferArrayPinned[0].Length = _count;
-                    wsaRecvMsgWSABufferArray = _wsaRecvMsgWSABufferArrayPinned;
-                    wsaRecvMsgWSABufferCount = 1;
+                    WSABuffer fixedBuffers = default;
 
-                    return Core();
+                    fixedBuffers.Pointer = (IntPtr)bufferPtr + _offset;
+                    fixedBuffers.Length = _count;
+
+                    return Core(&fixedBuffers, 1);
                 }
             }
             else
             {
                 // Use the multi-buffer WSABuffer.
-                wsaRecvMsgWSABufferArray = _wsaBufferArrayPinned!;
-                wsaRecvMsgWSABufferCount = (uint)_bufferListInternal!.Count;
+                int count = _bufferListInternal!.Count;
+                Span<WSABuffer> wsaBuffers = count <= MaxStackWSABuffers ? stackalloc WSABuffer[count] : GC.AllocateUninitializedArray<WSABuffer>(count, pinned: true);
+                SetupWsaBuffers(wsaBuffers);
 
-                return Core();
+                fixed (WSABuffer* buffersPtr = wsaBuffers)
```

Comment:
With this fixed, we don't need the `pinned` above, correct?  Which means the fallback can also just be `new WSABuffer[count]`?

Reply:
`ArrayPool.Rent(count)` would be even better, but it would complicate the codez.

Reply:
Yes, if its the allocation on the pinned heap that bothers you most about https://github.com/dotnet/runtime/pull/51078#discussion_r611230704 ?

Reply:
> if its the allocation on the pinned heap that bothers you most

Yes

Reply:
+1, temporary buffers on the pinned heap seem like a bad idea.